src/ext/java/org/opentripplanner/ext/fares/impl/HSLFareServiceImpl.java

```diff
@@ -161,18 +163,22 @@ but visit temporarily (maybe 1 stop only) an 'external' zone */
               );
             }
           }
-          float newFare = getFarePrice(attribute, fareType);
-          if (newFare < bestFare) {
+          Money newFare = getFarePrice(attribute, fareType);
+          if (newFare.lessThan(bestFare)) {
             bestAttribute = attribute;
             bestFare = newFare;
           }
         }
       }
-    } else if (specialRouteFare != Float.POSITIVE_INFINITY && specialFareAttribute != null) {
+    } else if (!specialRouteFare.equals(MAX_PRICE) && specialFareAttribute != null) {
       bestFare = specialRouteFare;
       bestAttribute = specialFareAttribute;
     }
     LOG.debug("HSL {} best for {}", bestAttribute, legs);
-    return new FareAndId(bestFare, bestAttribute == null ? null : bestAttribute.getId());
+    final FareAndId value = new FareAndId(
```

Comment:
Code does not capture the case when ticket was not found. This will cause NPE. 

Fix is simple: add this test before returning the value:

 ```
   if (bestAttribute == null) {
       return Optional.empty();
    }

```

Reply:
Oh, what an embarrassing oversight. I fixed it.